### PR TITLE
Update to CDI 4.1.0.Alpha1

### LINKED ISSUE #26791: Pull in CDI API 4.1 Alpha jars
CDI 4.1 now has Alpha releases available in Maven Central.

CDI API - 4.1.0.Alpha1
CDI TCK - 4.1.0.Alpha5

For those of you who want to play around with it, there is also Weld 6.0.0.Alpha1 implementing this new API - http://weld.cdi-spec.org/news/2023/10/26/weld-600Alpha1/

This includes implementation of the Method invokers API through MethodHandles.

### PR DESCRIPTION
**Pull request by tevans78**
- Update CDI 4.1 to use the 4.1.0.Alpha1 API and the 6.0.0.Alpha1 version of weld
- Remove the exact version number from the 2.x weld bundle symbolic name
- Provide an internal interface to be used by security to unwrap proxy object, avoiding their need to depend directly on weld.
- Rename the internal `com.ibm.ws.cdi40.extension` package to `io.openliberty.cdi40.internal.extension`
- Mark CDI 4.1 specific bundles as requiring Java 21
  - Note that `io.openliberty.org.jboss.weld6` is still set to Java 17 because our gradle builds don't use Java 21 yet for unit tests.
- Update CDI Core FAT to repeat with EE11 (CDI 4.1)

fixes https://github.com/OpenLiberty/open-liberty/issues/26791

### CODE REVIEW

**Comment by tevans78** [comment]
#build

File: dev/io.openliberty.cdi.4.0.internal.interfaces/src/io/openliberty/cdi40/internal/utils/CDI40Utils.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+     *
+     * @param instance The instance to unwrap
+     * @return The underlying instance
+     * @deprecated
+     */
+    @Deprecated
+    public static Object getContextualInstanceFromProxy(Object instance) {
+        Object contextualInstance = instance;
+        if (instance instanceof WeldClientProxy) {
+            contextualInstance = ((WeldClientProxy) instance).getMetadata().getContextualInstance();
```
Comment (tevans78) [inline dev/io.openliberty.cdi.4.0.internal.interfaces/src/io/openliberty/cdi40/internal/utils/CDI40Utils.java:33]: Does this need a null check?

File: dev/io.openliberty.jakarta/cdi.4.1.bnd [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+# 
+# SPDX-License-Identifier: EPL-2.0
+#*******************************************************************************
+-include= ~../cnf/resources/bnd/bundle.props
+bVersion=1.0
+
+javac.source: 11
+javac.target: 11
+
+Require-Capability: osgi.ee; filter:="(&(osgi.ee=JavaSE)(version=11))"
```
Comment (tevans78) [inline dev/io.openliberty.jakarta/cdi.4.1.bnd]: Should this be Java 21?

File: dev/io.openliberty.jakarta/interceptor.2.2.bnd [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+# 
+# SPDX-License-Identifier: EPL-2.0
+#*******************************************************************************
+-include= ~../cnf/resources/bnd/bundle.props
+bVersion=1.0
+
+javac.source: 11
+javac.target: 11
+
+Require-Capability: osgi.ee; filter:="(&(osgi.ee=JavaSE)(version=11))"
```
Comment (tevans78) [inline dev/io.openliberty.jakarta/interceptor.2.2.bnd]: Java 21?

File: dev/io.openliberty.org.jboss.weld6/bnd.template [hunk 4]
```diff
@@ -0,0 +1,48 @@
+###################################################################################
```
Comment (tevans78) [inline dev/io.openliberty.org.jboss.weld6/bnd.template:10]: Template should have a copyright header too

File: dev/io.openliberty.org.jboss.weld6/build.gradle [hunk 5]
```diff
@@ -0,0 +1,199 @@
+import org.gradle.api.java.archives.internal.DefaultManifest
```
Comment (tevans78) [inline dev/io.openliberty.org.jboss.weld6/build.gradle:1]: copyright header needed

File: dev/io.openliberty.org.jboss.weld6/check-me-in-if-i-change.bnd [hunk 6]
```diff
@@ -0,0 +1,228 @@
+###################################################################################
```
Comment (tevans78) [inline dev/io.openliberty.org.jboss.weld6/check-me-in-if-i-change.bnd:10]: Copyright header should be added via template

File: dev/io.openliberty.org.jboss.weld6/test/io/openliberty/weld/jsf/WeldJSFCheckSumTest.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+ * These tests look for changes in the FacesUrlTransformer & ConversationAwareViewHandler classes.
+ * If the checksum changes, then the latest classes should investigated and ported over to com.ibm.ws.jsfContainer
+ * See issue https://github.com/OpenLiberty/open-liberty/issues/14524
+ */
+public class WeldJSFCheckSumTest {
+
+    File bundle;
+
+    //TODO: this will all need to be updated once the EE10 version of JSF has been done
+    String bundleFileName = "io.openliberty.org.jboss.weld5.jar";
```
Comment (tevans78) [inline dev/io.openliberty.org.jboss.weld6/test/io/openliberty/weld/jsf/WeldJSFCheckSumTest.java]: This should say weld6 ... but then maybe the previous version needs updating anyway, as suggested by the comment?

**Comment by tevans78** [comment]
#build

**Comment by tevans78** [comment]
New PB running more CDI FAT buckets in FULL mode to include EE11 (CDI 4.1) testing.
https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_Y9k3YIhdEe6PKt0UtTmx2A

**Comment by tevans78** [comment]
#build

**Comment by tevans78** [comment]
#build

**Comment by tevans78** [comment]
#build

**Review by benjamin-confino — APPROVED** [review]

**Comment by tevans78** [comment]
#libby

**Pull request merged**